### ISSUE TITLE
Quantifiers in invariant

### ISSUE THREAD

**Issue by adilsonv77** [OpenJML/OpenJML#533]
Is it possible using quantifiers in invariant ?

Class to test sum quantifier in invariant.

```
public class TestSum {
	/*@
	 private invariant (\sum Integer i; list.contains(i); i) > 0; 
	 @*/
	private List<Integer> list = new ArrayList<Integer>();
	
	public void add(int i) {
		list.add(i);
	}
	}
}
```
Console output:
```
Internal JML bug - please report.  BuildOpenJML-20170527
java.lang.NullPointerException
	at com.sun.tools.javac.comp.Attr.enclosingInitEnv(Attr.java:3696)
	at com.sun.tools.javac.comp.JmlAttr.checkInit(JmlAttr.java:2025)
	at com.sun.tools.javac.comp.Attr.visitIdent(Attr.java:3204)
	at com.sun.tools.javac.comp.JmlAttr.visitIdent(JmlAttr.java:4610)
	at com.sun.tools.javac.tree.JCTree$JCIdent.accept(JCTree.java:2011)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.visitSelect(Attr.java:3252)
	at com.sun.tools.javac.comp.JmlAttr.visitSelect(JmlAttr.java:4945)
	at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.visitApply(Attr.java:1826)
	at com.sun.tools.javac.comp.JmlAttr.visitApply(JmlAttr.java:3544)
	at com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1465)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribExpr(Attr.java:618)
	at com.sun.tools.javac.comp.JmlAttr.attribExpr(JmlAttr.java:5889)
	at com.sun.tools.javac.comp.JmlAttr.visitJmlQuantifiedExpr(JmlAttr.java:4015)
	at org.jmlspecs.openjml.JmlTree$JmlQuantifiedExpr.accept(JmlTree.java:2461)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.visitParens(Attr.java:2987)
	at com.sun.tools.javac.tree.JCTree$JCParens.accept(JCTree.java:1661)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribExpr(Attr.java:625)
	at com.sun.tools.javac.comp.Attr.visitBinary(Attr.java:3057)
	at com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:1785)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribExpr(Attr.java:618)
	at com.sun.tools.javac.comp.JmlAttr.attribExpr(JmlAttr.java:5889)
	at com.sun.tools.javac.comp.JmlAttr.visitJmlTypeClauseExpr(JmlAttr.java:2415)
	at org.jmlspecs.openjml.JmlTree$JmlTypeClauseExpr.accept(JmlTree.java:3330)
	at com.sun.tools.javac.comp.JmlAttr.attribClassBodySpecs(JmlAttr.java:592)
	at com.sun.tools.javac.comp.JmlAttr.attribClassBody(JmlAttr.java:545)
	at com.sun.tools.javac.comp.Attr.attribClass(Attr.java:4274)
	at com.sun.tools.javac.comp.JmlAttr.attribClass(JmlAttr.java:423)
	at com.sun.tools.javac.comp.Attr.attribClass(Attr.java:4203)
	at com.sun.tools.javac.comp.Attr.attrib(Attr.java:4178)
	at com.sun.tools.javac.main.JavaCompiler.attribute(JavaCompiler.java:1258)
	at org.jmlspecs.openjml.JmlCompiler.attribute(JmlCompiler.java:425)
	at com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:898)
	at org.jmlspecs.openjml.JmlCompiler.compile2(JmlCompiler.java:640)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:867)
	at com.sun.tools.javac.main.Main.compile(Main.java:553)
	at com.sun.tools.javac.main.Main.compile(Main.java:410)
	at org.jmlspecs.openjml.Main.compile(Main.java:557)
	at com.sun.tools.javac.main.Main.compile(Main.java:399)
	at com.sun.tools.javac.main.Main.compile(Main.java:390)
	at org.jmlspecs.openjml.Main.executeNS(Main.java:483)
	at org.jmlspecs.openjml.API.execute(API.java:212)
	at org.jmlspecs.openjml.eclipse.OpenJMLInterface.executeExternalCommand(OpenJMLInterface.java:280)
	at org.jmlspecs.openjml.eclipse.JMLBuilder.doAction(JMLBuilder.java:198)
	at org.jmlspecs.openjml.eclipse.JMLBuilder.incrementalBuild(JMLBuilder.java:283)
	at org.jmlspecs.openjml.eclipse.JMLBuilder.build(JMLBuilder.java:131)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:734)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:205)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:245)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:300)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:303)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:359)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:382)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:144)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:235)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
ENDING with exit code 4
INTERNAL ERROR: return code = 4 
```

**Comment by davidcok** [comment]
In principle yes, but implementation of using Integer and of using contains is not completed (sorry - lambdas have had higher priority).

But the workaround, if you are doing RAC, is to write a pure function that computes the sum and then use that function in the invariant, as in

/*@ pure */ private Integer listsum() { … }

//@ private invariant listsum() > 0;

- David

**Comment by adilsonv77** [comment]
OK... The same happens when we dont use contains

```
        /*@
	 private invariant (\sum int i; i <= 0 && i < list.size(); list.get(i)) > 0; 
	 @*/
```

**Comment by davidcok** [comment]
Yes - that bug is fixed for 0.8.20 - but not yet the use of Integer or contains.

- David

**Comment by davidcok** [comment]
Fixed in V0.8.20
